c/src/ml-api-service-common.c

```diff
@@ -13,37 +13,84 @@
 #include "ml-api-service.h"
 #include "ml-api-service-private.h"
 
+#define ML_SERVICE_MAGIC 0xfeeedeed
+#define ML_SERVICE_MAGIC_DEAD 0xdeaddead
+
 /**
- * @brief Destroy the service handle.
+ * @brief Internal function to validate ml-service handle.
+ */
+gboolean
+_ml_service_handle_is_valid (ml_service_s * mls)
+{
+  if (!mls)
+    return FALSE;
+
+  if (mls->magic != ML_SERVICE_MAGIC)
+    return FALSE;
+
+  return TRUE;
+}
+
+/**
+ * @brief Internal function to create new ml-service handle.
+ */
+ml_service_s *
+_ml_service_create_internal (ml_service_type_e ml_service_type)
+{
+  ml_service_s *mls;
+
+  mls = g_try_new0 (ml_service_s, 1);
+  if (mls) {
+    mls->magic = ML_SERVICE_MAGIC;
+    mls->type = ml_service_type;
+  }
+
+  return mls;
+}
+
+/**
+ * @brief Internal function to release ml-service handle.
  */
 int
-ml_service_destroy (ml_service_h handle)
+_ml_service_destroy_internal (ml_service_s * mls)
 {
   int ret = ML_ERROR_NONE;
-  ml_service_s *mls = (ml_service_s *) handle;
-
-  check_feature_state (ML_FEATURE_SERVICE);
 
-  if (!handle)
+  if (!_ml_service_handle_is_valid (mls)) {
     _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
-        "The parameter, 'handle' is NULL. It should be a valid ml_service_h.");
+        "The parameter, 'handle' (ml_service_h), is invalid. It should be a valid ml_service_h instance.");
+  }
 
   switch (mls->type) {
     case ML_SERVICE_TYPE_SERVER_PIPELINE:
-      ret = ml_service_pipeline_release_internal (mls->priv);
+      ret = ml_service_pipeline_release_internal (mls);
       break;
     case ML_SERVICE_TYPE_CLIENT_QUERY:
-      ret = ml_service_query_release_internal (mls->priv);
+      ret = ml_service_query_release_internal (mls);
       break;
     case ML_SERVICE_TYPE_REMOTE:
-      ret = ml_service_remote_release_internal (mls->priv);
+      ret = ml_service_remote_release_internal (mls);
```

Comment:
When calling ```_ml_service_destroy_internal()```, when ret is an error, I wonder if there is a need to handle the error in the part where the function is called.

Reply:
I will add log messge if ret != ML_ERROR_NONR.

Reply:
updated